lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of it's contents from the nfs store. We already know via the cache if it is in use.
+func (v *NFSv3VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+
+	defer v.MH.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and it's metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove volume (%s) on volume store (%s) due to error (%s)", vol.ID, v.Name, err)
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
+
+	//remove volume metadata directory and children
+	if err := target.RemoveAll(v.volMetadataDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove metadata for volume (%s) at path (%q) on volume store (%s)", vol.ID, v.volDirPath(vol.ID), v.Name)
+		//FIXME: Should we bail here? the volume is gone at this point...
+	}
+	op.Infof("Successfully removed volume (%s) from volumestore (%s)", vol.ID, v.Name)
+
+	return nil
+}
+
+func (v *NFSv3VolumeStore) VolumesList(op trace.Operation) ([]*storage.Volume, error) {
+	volumes := make([]*storage.Volume, 0)
+
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	volFileInfo, err := target.ReadDir(v.volumesDir())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volFileInfo := range volFileInfo {
+
+		volMetadata, err := getMetadata(op, v.volMetadataDirPath(volFileInfo.Name()), target)
+		if err != nil {
+			return nil, err //perhaps wait until the end of the loop to report this?
+		}
+
+		volDeviceBacking := NewNFSVolumeDevice(v.Target, v.volDirPath(volFileInfo.Name()))
+
+		vol, err := storage.NewVolume(v.SelfLink, volFileInfo.Name(), volMetadata, volDeviceBacking)
+		if err != nil {
+			return nil, err
+		}
+
+		volumes = append(volumes, vol)
+	}
+
+	return volumes, nil
+}
+
+func writeMetadata(op trace.Operation, metadataPath string, info map[string][]byte, target NFSTarget) error {
+	//NOTE: right now we do not support updating metadata, thus we make the ID directory here
+	_, err := target.MkDir(metadataPath, filePermissions)
+	if err != nil {
+		return err
+	}
+
+	op.Infof("Attempting to write metadata to (%s)", metadataPath)
+	for fileName, data := range info {
+		targetPath := path.Join(metadataPath, fileName)
+		blobFile, err := target.Create(targetPath, filePermissions)
+		// XXX: we might want to make sure the file we want to write does not exist...
+		if err != nil {
+			return err
+		}
+
+		_, err = blobFile.Write(data)
+		if err != nil {
+			//XXX: Should we attempt to clean up what we wrote in this case?
```

Comment:
Same consistency comment;  we'll revisit what to do in this case.  Likely for this specific case, we'll clean up in the cache.